zerver/lib/export.py

```diff
@@ -0,0 +1,653 @@
+    bucket = conn.get_bucket(bucket_name, validate=True)
+    records = []
+
+    logging.info("Downloading uploaded files from %s" % (bucket_name))
+
+    avatar_hash_values = set()
+    user_ids = set()
+    if avatar_bucket:
+        bucket_list = bucket.list()
+        for user_profile in UserProfile.objects.filter(realm=realm):
+            avatar_hash = user_avatar_hash(user_profile.email)
+            avatar_hash_values.add(avatar_hash)
+            avatar_hash_values.add(avatar_hash + ".original")
+            user_ids.add(user_profile.id)
+    else:
+        bucket_list = bucket.list(prefix="%s/" % (realm.id,))
+
+    if not os.path.exists(output_dir):
+        os.makedirs(output_dir)
+
+    if settings.EMAIL_GATEWAY_BOT is not None:
+        email_gateway_bot = get_user_profile_by_email(settings.EMAIL_GATEWAY_BOT)
+    else:
+        email_gateway_bot = None
+
+    count = 0
+    for bkey in bucket_list:
+        if avatar_bucket and bkey.name not in avatar_hash_values:
+            continue
+        key = bucket.get_key(bkey.name)
+
+        # This can happen if an email address has moved realms
+        if 'realm_id' in key.metadata and key.metadata['realm_id'] != str(realm.id):
+            if email_gateway_bot is None or key.metadata['user_profile_id'] != str(email_gateway_bot.id):
+                raise Exception("Key metadata problem: %s %s / %s" % (key.name, key.metadata, realm.id))
+            # Email gateway bot sends messages, potentially including attachments, cross-realm.
+            print("File uploaded by email gateway bot: %s / %s" % (key.name, key.metadata))
+        elif avatar_bucket:
+            if 'user_profile_id' not in key.metadata:
+                raise Exception("Missing user_profile_id in key metadata: %s" % (key.metadata,))
+            if int(key.metadata['user_profile_id']) not in user_ids:
+                raise Exception("Wrong user_profile_id in key metadata: %s" % (key.metadata,))
+        elif 'realm_id' not in key.metadata:
+            raise Exception("Missing realm_id in key metadata: %s" % (key.metadata,))
+
+        record = dict(s3_path=key.name, bucket=bucket_name,
+                      size=key.size, last_modified=key.last_modified,
+                      content_type=key.content_type, md5=key.md5)
+        record.update(key.metadata)
+
+        # A few early avatars don't have 'realm_id' on the object; fix their metadata
+        user_profile = get_user_profile_by_id(record['user_profile_id'])
+        if 'realm_id' not in record:
+            record['realm_id'] = user_profile.realm_id
+        record['user_profile_email'] = user_profile.email
+
+        if avatar_bucket:
+            dirname = output_dir
+            filename = os.path.join(dirname, key.name)
+            record['path'] = key.name
+        else:
+            fields = key.name.split('/')
+            if len(fields) != 3:
+                raise Exception("Suspicious key %s" % (key.name))
```

Comment:
As I mentioned on Zulip, we probably want to make sure S3 stuff doesn't derail the rest of the migration, so here it might make sense to log/continue instead of raising an error.